descartes_trajectory/src/cart_trajectory_pt.cpp

```diff
@@ -184,29 +184,60 @@ bool CartTrajectoryPt::getNominalCartPose(const std::vector<double> &seed_state,
   return true;  //TODO can this ever return false?
 }
 
-void CartTrajectoryPt::getCartesianPoses(const RobotModel &model, EigenSTL::vector_Affine3d &poses) const
+bool CartTrajectoryPt::computeCartesianPoses(EigenSTL::vector_Affine3d &poses) const
 {
-  poses.clear();
   EigenSTL::vector_Affine3d sampled_wobj_pts = uniform(wobj_pt_, orient_increment_,
                                                      pos_increment_);
-  poses.reserve(sampled_wobj_pts.size());
+  EigenSTL::vector_Affine3d sampled_tool_pts = uniform(tool_pt_, orient_increment_,
+                                                       pos_increment_);
+
+  poses.clear();
+  poses.reserve(sampled_wobj_pts.size()*sampled_tool_pts.size());
   for(size_t wobj_pt = 0; wobj_pt < sampled_wobj_pts.size(); ++wobj_pt)
   {
-    Eigen::Affine3d pose = wobj_base_.frame * sampled_wobj_pts[wobj_pt];
-    if(model.isValid(pose))
+    for(size_t tool_pt = 0; tool_pt < sampled_tool_pts.size(); ++tool_pt)
     {
+      Eigen::Affine3d pose = wobj_base_.frame * sampled_wobj_pts[wobj_pt] *
+          sampled_tool_pts[tool_pt].inverse() * tool_base_.frame_inv;
+
       poses.push_back(pose);
     }
   }
+
+  return !poses.empty();
+}
+
+void CartTrajectoryPt::getCartesianPoses(const RobotModel &model, EigenSTL::vector_Affine3d &poses) const
+{
+  EigenSTL::vector_Affine3d all_poses;
+  poses.clear();
+
+  if(computeCartesianPoses(all_poses))
+  {
+    poses.reserve(all_poses.size());
+    for(auto pose: all_poses)
+    {
+      if(model.isValid(pose))
+      {
+        poses.push_back(pose);
```

Comment:
Since we're doing the C++11 thing, situations like this are great opportunities for using move semantics. Here transfering ownership of the object state is preferrable to copying. I don't know if eigen supports move constructors, but food for thought.